core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -340,39 +338,48 @@ private[spark] class TaskSchedulerImpl(
     for (i <- 0 until shuffledOffers.size) {
       val execId = shuffledOffers(i).executorId
       val host = shuffledOffers(i).host
-      if (availableCpus(i) >= CPUS_PER_TASK &&
-        resourcesMeetTaskRequirements(availableResources(i))) {
-        try {
-          for (task <- taskSet.resourceOffer(execId, host, maxLocality, availableResources(i))) {
-            tasks(i) += task
-            val tid = task.taskId
-            taskIdToTaskSetManager.put(tid, taskSet)
-            taskIdToExecutorId(tid) = execId
-            executorIdToRunningTaskIds(execId).add(tid)
-            availableCpus(i) -= CPUS_PER_TASK
-            assert(availableCpus(i) >= 0)
-            task.resources.foreach { case (rName, rInfo) =>
-              // Remove the first n elements from availableResources addresses, these removed
-              // addresses are the same as that we allocated in taskSet.resourceOffer() since it's
-              // synchronized. We don't remove the exact addresses allocated because the current
-              // approach produces the identical result with less time complexity.
-              availableResources(i).getOrElse(rName,
-                throw new SparkException(s"Try to acquire resource $rName that doesn't exist."))
-                .remove(0, rInfo.addresses.size)
-            }
-            // Only update hosts for a barrier task.
-            if (taskSet.isBarrier) {
-              // The executor address is expected to be non empty.
-              addressesWithDescs += (shuffledOffers(i).address.get -> task)
+      val taskSetRpID = taskSet.taskSet.resourceProfileId
+      // make the resource profile id a hard requirement for now - ie only put tasksets
+      // on executors where resource profile exactly matches.
+      if (taskSetRpID == shuffledOffers(i).resourceProfileId) {
+        val taskResAssignmentsOpt = resourcesMeetTaskRequirements(taskSet, availableCpus(i),
+          availableResources(i))
+        taskResAssignmentsOpt.foreach { taskResAssignments =>
+          try {
+            val taskCpus = sc.resourceProfileManager.taskCpusForProfileId(taskSetRpID)
+            val taskDescOption = taskSet.resourceOffer(execId, host, maxLocality, taskCpus,
+              taskResAssignments)
+            for (task <- taskDescOption) {
+              tasks(i) += task
+              val tid = task.taskId
+              taskIdToTaskSetManager.put(tid, taskSet)
+              taskIdToExecutorId(tid) = execId
+              executorIdToRunningTaskIds(execId).add(tid)
+              availableCpus(i) -= task.cpus
+              assert(availableCpus(i) >= 0)
```

Comment:
Seem this is not necessary anymore since we've checked in `resourcesMeetTaskRequirements`. But I'm OK to keep it, though.

Reply:
this was always the case - even when cpu only, we left it when adding the original resource scheduling code as it doesn't hurt and just double check if something really unexpected happens